symengine/functions.cpp

```diff
@@ -3269,6 +3269,49 @@ RCP<const Basic> trigamma(const RCP<const Basic> &x)
     return polygamma(one, x);
 }
 
+Factorial::Factorial(const RCP<const Basic> &s) : OneArgFunction(s)
+{
+    SYMENGINE_ASSIGN_TYPEID()
+    SYMENGINE_ASSERT(is_canonical(s))
+}
+
+bool Factorial::is_canonical(const RCP<const Basic> &arg) const
+{
+    if (is_a<NaN>(*arg)) {
+        return false;
+    } else if (is_a<Infty>(*arg)) {
+        if (down_cast<const Infty &>(*arg).is_positive()) {
+            return false;
+        }
+    } else if (is_a<Integer>(*arg)
+               or (is_a_Number(*arg)
+                   and not down_cast<const Number &>(*arg).is_exact())) {
+        return false;
+    }
+    return true;
+}
+
+RCP<const Basic> Factorial::create(const RCP<const Basic> &arg) const
+{
+    return factorial(arg);
+}
+
+RCP<const Basic> factorial(const RCP<const Basic> &arg)
+{
+    if (is_a<NaN>(*arg)) {
+        return Nan;
+    } else if (is_a<Infty>(*arg)) {
+        if (down_cast<const Infty &>(*arg).is_positive()) {
+            return Inf;
+        }
+    } else if (is_a<Integer>(*arg)
+               or (is_a_Number(*arg)
+                   and not down_cast<const Number &>(*arg).is_exact())) {
+        return gamma(add(arg, one));
```

Comment:
There's a `factorial` in `ntheory` that you can use here.

Reply:
My concern is regarding code duplication that would occur. `gamma` already has the cases defined. It would be a literal copy-paste.

Reply:
Okay.

Reply:
This returns an instance of `Gamma` sometimes. Can you return a `Factorial` instance in that case?